### DISCUSSION TITLE
Managing multiple organizations

### DISCUSSION THREAD

**Discussion by lmilbaum** [integrations/terraform-provider-github#1923] [Q&A]
i'd like to manage multiple organizations. How do I go about it without duplicating the code?

**Comment by dion-gionet** [comment]
If you want to use mostly the same code you can use Terraform workspaces with tfvars to configure the github provider `owner` per organization

**Reply by lmilbaum** [reply]
Thanks @dion-gionet for your suggestion.  This option was considered.  Eventually I have used the `alias` feature.

**Reply by jamesgeddes** [reply]
Just as an update, while workspaces were commonly suggested in 2023, HashiCorp now generally [discourages](https://youtu.be/6QgHLncP5VA) their use outside very simple cases.

**Reply by lmilbaum** [reply]
Thanks for the info. I would guess that terragrunt could be of value here.

**Comment by testworksau** [comment]
I'm quite curious about this too; we are on GitHub Enterprise Cloud (GHEC) and would like to set up different organisations underneath the Enterprise, each with a common set of teams (eg. infosec and admins) and other teams that would only exist in a given org.   

I'd like to have one deployment repo with the relevant, shared terraform inside. 

I can create multiple `resource "github_enterprise_organization"` blocks but other resources like teams don't seem to be tied to a given organisation. 

Is the recommended approach to structure the code with an alias per organisation and a module per github provider alias?

**Reply by testworksau** [reply]
@lmilbaum would you mind sharing a little more about how you tackled this, & how you landed on `alias` being a better fit?

**Reply by veronluc-ansys** [reply]
@testworksau I am curious. Is `github_enterprise_organization` compatible with github enterprise cloud? Did you manage to make organization management with it in the end?

**Reply by testworksau** [reply]
Hi @veronluc-ansys 👋🏼 

Yes, it's compatible with GHEC however it doesn't work particularly well:

https://github.com/integrations/terraform-provider-github/issues/1695

We ended up ditching the usage of this resource, opting instead to create our organisations manually, and then adding modules for each organisation with their own `github_organization_settings` and other entities. 

It's a bit of a mess and there's likely a better way, but here's what our config roughly looks like.

provider.tf
```
terraform {
  required_providers {
    github = {
      source  = "integrations/github"
      version = "~> 6.2.0"
    }
  }
}

# The default provider, which is only used to get the enterprise id
provider "github" {
  token = var.github_access_token
}

# Each module requires its own provider definition to be passed in at
# module initialisation, due to a clash in the provider (it used to be
# in the Hashicorp namespace but is now in the owned by GitHub)
provider "github" {
  alias = "our_sandpit1"
  owner = "our-sandpit1"
  token = var.github_access_token
}

provider "github" {
  alias = "our_sandpit2"
  owner = "our-sandpit2"
  token = var.github_access_token
}
```

main.tf
```
module "our-sandpit1" {
  source = "./modules/org-our-sandpit1"

  providers = {
    github = github.our_sandpit1
  }

  github_access_token       = var.github_access_token
  github_org_administrators = var.github_default_administrators
  github_org_billing_email  = var.github_billing_email
}

module "our-sandpit2" {
  source = "./modules/org-our-sandpit2"

  providers = {
    github = github.our_sandpit2
  }

  github_access_token       = var.github_access_token
  github_org_administrators = var.github_default_administrators
  github_org_billing_email  = var.github_billing_email
}
```

We have a common github settings module for admins, org security managers etc. that we defined under:
modules/common_github_settings/

And then we have the org specific settings defined under modules/org-<name>.

Each module we have redefines the provider.tf:

```
# When a child module uses a provider outside of the Hashicorp namespace,
# you are required to duplicate the the entire terraform { required_providers { } }
# code block in every module where that provider is used, regardless if it’s already
# in the root module’s main.tf file.

terraform {
  required_providers {
    github = {
      source  = "integrations/github"
      version = "~> 6.2.0"
    }
  }
}
```

The `main.tf` in each org specific module loads the `common_github_settings` module:

```
module "common_github_settings" {
  source = "../../modules/common_github_settings"

  providers = {
    github = github
  }

  github_enterprise_organization = local.organization_name
  github_access_token            = var.github_access_token
  github_administrators          = var.github_org_administrators
}
```

Finally each module uses the github tf provider resources etc. as desired.

Hope this helps you / someone as it was a bit of a painful experience getting this sorted.

**Comment by jamesgeddes** [comment]
This is jolly interesting! 

@lmilbaum I know I’m a _little_ late to the conversation, but could you elaborate on your need for multiple GitHub orgs? For example, are you running a separate org per business unit?

**Reply by lmilbaum** [reply]
Yes. That sounds right.

**Reply by testworksau** [reply]
We are too; well, we have 2 per business unit - one for production / real code, another for experimentation with less restrictions.

**Reply by jamesgeddes** [reply]
Although OP’s problem is likely long since resolved, I haven’t yet run into this situation myself, so I am somewhat [nerd sniped](https://xkcd.com/356/). From what I can tell, `github_enterprise_organization` can both create and manage enterprise orgs, whereas `github_organization_settings` only manages existing orgs. If I had seen this in 2023, I might have suggested creating a module with sensible defaults and then `for_each` over it with a map of enterprise organisations.


```hcl
locals {
  enterprise_organizations = {
    "Reliability Guild" = {
      billing_email = "reliability@example.com"
      description   = "SRE and reliability engineering"
      members_allowed_repository_creation_type = "none" # stricter than default
    }

    "Foundation Services" = {
      billing_email = "foundation@example.com"
      description   = "Core shared services"
      ip_allowlist  = ["203.0.113.0/24"] # only this org needs an allowlist
    }

    "The Kernel Group" = {
      billing_email = "kernel@example.com"
      description   = "Low-level systems"
      default_repository_permission = "admin" # unusual for small trusted group
    }
  }
}

module "orgs" {
  source   = "[modules repo]/github_organization"
  for_each = local.enterprise_organizations

  name      = each.key
  overrides = each.value
}
```

When creating modules for the primary purpose of setting sensible defaults, `nullable = false` is your friend as it forces the use of the module default. [docs](https://developer.hashicorp.com/terraform/language/values/variables#disallowing-null-input-values)

Within that github_organization module, one would use `github_enterprise_organization` to create the org and then `github_organization_settings` to set the majority of the settings. The `github_organization_settings` seems to require a `depends_on` on the `github_enterprise_organization` else conflicts can occur.

I hope that might be of use to someone at some point, even if not for OP. Thanks for giving me a reason to dig into that 😂

**Reply by testworksau** [reply]
Did you hit this issue in your travels @jamesgeddes? 
https://github.com/integrations/terraform-provider-github/issues/1695

**Reply by jamesgeddes** [reply]
I recently hit a very similar pattern but not this specific matter relating to multiple orgs. Apologies if my suggestion is not quite right.
